File: 063_Subsets.cpp
// Topic: Subsets
// Description: Generate all possible subsets of a list.
// Approach: Backtracking or iterative bitmask. We'll use backtracking.

#include <bits/stdc++.h>
using namespace std;

class Solution
{
   public:
    vector<vector<int>> subsets(vector<int>& nums)
    {
        vector<vector<int>> res;
        vector<int> cur;
        function<void(int)> dfs = [&](int idx)
        {
            res.push_back(cur);
            for (int i = idx; i < (int)nums.size(); ++i)
            {
                cur.push_back(nums[i]);
                dfs(i + 1);
                cur.pop_back();
            }
        };
        dfs(0);
        return res;
    }
};

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    vector<int> a = {1, 2, 3};
    auto res = Solution().subsets(a);
    for (auto& s : res)
    {
        for (int i = 0; i < (int)s.size(); ++i)
        {
            if (i)
                cout << " ";
            cout << s[i];
        }
        cout << "\n";
    }
}

/*
Time Complexity: O(n * 2^n) total output size
Space Complexity: O(n) recursion depth + output
*/

File: 063_Subsets.py
# Topic: Subsets
# Description: Generate all possible subsets of a list.
# Approach: Backtracking or iterative bitmask. We'll use backtracking.

from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []
        cur = []

        def dfs(idx):
            result.append(cur[:])
            for i in range(idx, len(nums)):
                cur.append(nums[i])
                dfs(i + 1)
                cur.pop()

        dfs(0)
        return result


if __name__ == "__main__":
    nums = [1, 2, 3]
    result = Solution().subsets(nums)
    for s in result:
        print(s)

"""
Time Complexity: O(n * 2^n) total output size
Space Complexity: O(n) recursion depth + output
"""
